mmdet/models/dense_heads/ddod_head.py

```diff
@@ -0,0 +1,641 @@
+                    levels, each is a 4D-tensor, the channels number is
+                    num_base_priors * 4.
+                iou_preds (list[Tensor]): IoU scores for all scale levels,
+                    each is a 4D-tensor, the channels number is
+                    num_base_priors * 1.
+        """
+        return multi_apply(self.forward_single, feats, self.scales)
+
+    def forward_single(self, x, scale):
+        """Forward feature of a single scale level.
+
+        Args:
+            x (Tensor): Features of a single scale level.
+            scale (:obj: `mmcv.cnn.Scale`): Learnable scale module to resize
+                the bbox prediction.
+
+        Returns:
+            tuple:
+                - cls_score (Tensor): Cls scores for a single scale level \
+                    the channels number is num_base_priors * num_classes.
+                - bbox_pred (Tensor): Box energies / deltas for a single \
+                    scale level, the channels number is num_base_priors * 4.
+                - iou (Tensor): Iou for a single scale level, the \
+                    channel number is (N, num_base_priors * 1, H, W).
+        """
+        cls_feat = x
+        reg_feat = x
+        for cls_conv in self.cls_convs:
+            cls_feat = cls_conv(cls_feat)
+        for reg_conv in self.reg_convs:
+            reg_feat = reg_conv(reg_feat)
+        cls_score = self.atss_cls(cls_feat)
+        # we just follow atss, not apply exp in bbox_pred
+        bbox_pred = scale(self.atss_reg(reg_feat)).float()
+        iou_pred = self.atss_iou(reg_feat)
+        return cls_score, bbox_pred, iou_pred
+
+    def loss_single(self, anchors, cls_score, bbox_pred, iou_pred, labels,
+                    label_weights, bbox_targets, bbox_weights, reweight_factor,
+                    num_total_samples, is_cls_assigner):
+        """Compute loss of a single scale level.
+
+        Args:
+            cls_score (Tensor): Box scores for each scale level
+                Has shape (N, num_base_priors * num_classes, H, W).
+            bbox_pred (Tensor): Box energies / deltas for each scale
+                level with shape (N, num_base_priors * 4, H, W).
+            anchors (Tensor): Box reference for each scale level with shape
+                (N, num_total_anchors, 4).
+            labels (Tensor): Labels of each anchors with shape
+                (N, num_total_anchors).
+            label_weights (Tensor): Label weights of each anchor with shape
+                (N, num_total_anchors)
+            bbox_targets (Tensor): BBox regression targets of each anchor
+                weight shape (N, num_total_anchors, 4).
+            num_total_samples (int): Number of positive samples that is
+                reduced over all GPUs.
+            is_cls_assigner (bool): Classification or regression.
+
+        Returns:
+            dict[str, Tensor]: A dictionary of loss components.
+        """
+        if is_cls_assigner:
```

Comment:
I suggest splitting the two parts of logic into two loss functions(like loss_cls_single&loss_reg_single) instead of taking two branches in one function.

Reply:
I will seperate these two loss functions from one part. Modifying.